File: 2025/day09/main.py
import sys
from functools import cache

def part1(input: str) -> str:
    red_tiles = [tuple(map(int, line.split(','))) for line in input.strip().split("\n")]
    
    biggest_area_rec = 0
    for i, t1 in enumerate(red_tiles):
        for t2 in red_tiles[i+1:]:
            biggest_area_rec = max(biggest_area_rec, (abs(t1[0] - t2[0]) + 1)*(abs(t1[1] - t2[1])+1))
    
    return str(biggest_area_rec)

# Deepseek ftw
def part2(input_str: str) -> str:
    red_tiles = [tuple(map(int, line.split(','))) for line in input_str.strip().split("\n")]
    n = len(red_tiles)
    
    polygon = red_tiles + [red_tiles[0]]
    
    def point_in_polygon(px, py):
        for i in range(n):
            x1, y1 = polygon[i]
            x2, y2 = polygon[i + 1]
            if x1 == x2 and px == x1 and min(y1, y2) <= py <= max(y1, y2):
                return True
            if y1 == y2 and py == y1 and min(x1, x2) <= px <= max(x1, x2):
                return True
        
        inside = False
        for i in range(n):
            x1, y1 = polygon[i]
            x2, y2 = polygon[i + 1]
            if ((y1 > py) != (y2 > py)) and (px < (x2 - x1) * (py - y1) / (y2 - y1) + x1):
                inside = not inside
        return inside
    
    xs = sorted({x for x, _ in red_tiles})
    ys = sorted({y for _, y in red_tiles})
    
    x_to_idx = {x: i for i, x in enumerate(xs)}
    y_to_idx = {y: i for i, y in enumerate(ys)}
    
    w = len(xs) - 1
    h = len(ys) - 1

    inside_grid = [[False] * h for _ in range(w)]
    for i in range(w):
        for j in range(h):
            cx = (xs[i] + xs[i + 1]) / 2
            cy = (ys[j] + ys[j + 1]) / 2
            inside_grid[i][j] = point_in_polygon(cx, cy)
    
    prefix_sum = [[0] * (h + 1) for _ in range(w + 1)]
    for i in range(w):
        for j in range(h):
            is_inside = 1 if inside_grid[i][j] else 0
            prefix_sum[i + 1][j + 1] = (prefix_sum[i + 1][j] + prefix_sum[i][j + 1] - prefix_sum[i][j] + is_inside)
    
    def calc_actual_area(x1, y1, x2, y2):
        return prefix_sum[x2][y2] - prefix_sum[x1][y2] - prefix_sum[x2][y1] + prefix_sum[x1][y1]
    
    biggest_area_rec = 0
    for i, (x1, y1) in enumerate(red_tiles):
        idx1_x, idx1_y = x_to_idx[x1], y_to_idx[y1]
        
        for x2, y2 in red_tiles:
            idx2_x, idx2_y = x_to_idx[x2], y_to_idx[y2]
            
            min_idx_x = min(idx1_x, idx2_x)
            max_idx_x = max(idx1_x, idx2_x)
            min_idx_y = min(idx1_y, idx2_y)
            max_idx_y = max(idx1_y, idx2_y)
                
            expected_area = (max_idx_x - min_idx_x) * (max_idx_y - min_idx_y)
            actual_area = calc_actual_area(min_idx_x, min_idx_y, max_idx_x, max_idx_y)
            if actual_area == expected_area:
                biggest_area_rec = max(biggest_area_rec, (abs(x1 - x2) + 1)*(abs(y1 - y2) + 1))
    
    return str(biggest_area_rec)


def main():
    if len(sys.argv) != 2: print("Usage: main.py input.txt"); exit(1)
    with open(sys.argv[1]) as infile: input = infile.read()
    print("Part 1:", part1(input))
    print("Part 2:", part2(input))

if __name__ == "__main__":
    main()

File: 2025/c/day09/main.c
#include <assert.h>
#include <stddef.h>
#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#define NOB_IMPLEMENTATION
#define NOB_STRIP_PREFIX
#include "nob.h"

uint64_t svtou64(Nob_String_View input) {
    bool is_neg = input.data[0] == '-';
    uint64_t res = 0;
    for(size_t i = is_neg ? 1 : 0; i < input.count; i++) {
        if(input.data[i] < '0' || input.data[i] > '9') return 0;
        res = res * 10 + input.data[i] - '0';
    }
    return is_neg ? -res : res;
}

typedef struct {
    uint64_t x;
    uint64_t y;
} Point;

uint64_t abs_diff(uint64_t a, uint64_t b) {
    if(a > b) return a - b;
    return b - a;
}

uint64_t rect_area(Point t1, Point t2) {
    return (abs_diff(t1.x, t2.x) + 1)*(abs_diff(t1.y, t2.y) + 1);
}

uint64_t max(uint64_t a, uint64_t b) {
    if(a > b) return a;
    return b;
}

uint64_t min(uint64_t a, uint64_t b) {
    if(a < b) return a;
    return b;
}


typedef struct {
    Point *items;
    size_t count;
    size_t capacity;
} Points;

void part1(Nob_String_View input) {
    Points tiles = {0};
    while(input.count > 0) {
        Nob_String_View line = nob_sv_chop_by_delim(&input, '\n');
        Point tile = {
            svtou64(nob_sv_chop_by_delim(&line, ',')),
            svtou64(nob_sv_chop_by_delim(&line, ',')),
        };
        nob_da_append(&tiles, tile);
    }


    uint64_t bigest_area = 0;
    for(size_t i = 0; i < tiles.count; i++) {
        for(size_t j = i+1; j < tiles.count; j++) {
            bigest_area = max(bigest_area, rect_area(tiles.items[i], tiles.items[j]));
        }
    }

    printf("Part 1: %"PRId64"\n", bigest_area);
}
typedef struct{
    uint64_t *items;
    size_t count;
    size_t capacity;
} Ints;

typedef struct{
    bool *items;
    size_t count;
    size_t capacity;
} Bools;

int compare_uint64t(const void* v1v, const void* v2v) {
    size_t v1 = *(uint64_t*)v1v, v2 = *(uint64_t*)v2v;
    if(v1 == v2) return 0;
    if(v1 < v2) return -1;
    return 1;
}

int find_index(Ints ints, uint64_t v) {
    for(size_t i = 0; i < ints.count; i++)
        if(ints.items[i] == v) 
            return i;
    
    return -1;
}

bool point_in_polygon(Points *polygon, double px, double py) {
    for(size_t i = 0; i < polygon->count-1; i++) {
        Point p1 = polygon->items[i];
        Point p2 = polygon->items[i + 1];
        if((p1.x == p2.x) && (px == p1.x) && min(p1.y, p2.y) <= py && py <= max(p1.y, p2.y)) {
            return true;
        }
        if((p1.y == p2.y) && (py == p1.y) && min(p1.x, p2.x) <= px && px <= max(p1.x, p2.x)) {
            return true;
        }
    }

    bool inside = false;
    for(size_t i = 0; i < polygon->count-1; i++) {
        Point p1 = polygon->items[i];
        Point p2 = polygon->items[i + 1];
        if((p1.y > py) != (p2.y > py) && (px < (double)(p2.x - p1.x) * (py - p1.y) / (p2.y - p1.y) + p1.x)) {
            inside = !inside;
        }
    }

    return inside;
}

uint64_t calc_actual_area(Ints *prefix_sum, size_t w, uint64_t x1, uint64_t y1, uint64_t x2, uint64_t y2) {
    return prefix_sum->items[x2 + y2*(w + 1)] 
           - prefix_sum->items[x1 + y2*(w + 1)] 
           - prefix_sum->items[x2 + y1*(w + 1)] 
           + prefix_sum->items[x1 + y1*(w + 1)];
}

void part2(Nob_String_View input) {
    Points tiles = {0};
    Points polygon = {0};
    Ints xs = {0};
    Ints ys = {0};
    while(input.count > 0) {
        Nob_String_View line = nob_sv_chop_by_delim(&input, '\n');
        Point tile = {
            svtou64(nob_sv_chop_by_delim(&line, ',')),
            svtou64(nob_sv_chop_by_delim(&line, ',')),
        };
        nob_da_append(&tiles, tile);
        nob_da_append(&polygon, tile);
        if(find_index(xs, tile.x) == -1) nob_da_append(&xs, tile.x);
        if(find_index(ys, tile.y) == -1) nob_da_append(&ys, tile.y);
    }
    nob_da_append(&polygon, tiles.items[0]);
    qsort(xs.items, xs.count, sizeof(uint64_t), compare_uint64t);
    qsort(ys.items, ys.count, sizeof(uint64_t), compare_uint64t);
    
    size_t w = xs.count - 1;
    size_t h = ys.count - 1;

    Bools inside_grid = {0};
    for(size_t i = 0; i < w; i++) {
        for(size_t j = 0; j < h; j++) {
            nob_da_append(&inside_grid, false);
        }
    }
    for(size_t i = 0; i < w; i++) {
        for(size_t j = 0; j < h; j++) {
            double cx = (double)(xs.items[i] + xs.items[i + 1]) / 2;
            double cy = (double)(ys.items[j] + ys.items[j + 1]) / 2;
            inside_grid.items[i + j*w] = point_in_polygon(&polygon, cx, cy);
        }
    }

    Ints prefix_sum = {0};
    for(size_t i = 0; i < w + 1; i++) {
        for(size_t j = 0; j < h + 1; j++) {
            nob_da_append(&prefix_sum, 0);
        }
    }

    for (size_t i = 0; i < w; ++i) {
        for (size_t j = 0; j < h; ++j) {
            prefix_sum.items[(i + 1) + (j + 1) * (w + 1)] = 
                                    prefix_sum.items[(i + 1) + j * (w + 1)] 
                                    + prefix_sum.items[i + (j + 1) * (w + 1)] 
                                    - prefix_sum.items[i + j * (w + 1)] 
                                    + (inside_grid.items[i + j * w] ? 1 : 0);
        }
    }

    uint64_t biggest_area_rec = 0;
    for(size_t i = 0; i < tiles.count; i++) {
        Point tile = tiles.items[i];
        size_t idx1_x = find_index(xs, tile.x);
        size_t idx1_y = find_index(ys, tile.y);

        for(size_t j = 0; j < tiles.count; j++) {
            Point tile2 = tiles.items[j];
            size_t idx2_x = find_index(xs, tile2.x);
            size_t idx2_y = find_index(ys, tile2.y);

            uint64_t min_idx_x = min(idx1_x, idx2_x);
            uint64_t max_idx_x = max(idx1_x, idx2_x);
            uint64_t min_idx_y = min(idx1_y, idx2_y);
            uint64_t max_idx_y = max(idx1_y, idx2_y);

            uint64_t expected_area = (max_idx_x - min_idx_x) * (max_idx_y - min_idx_y);
            uint64_t actual_area = calc_actual_area(&prefix_sum, w, min_idx_x, min_idx_y, max_idx_x, max_idx_y);
            if(actual_area == expected_area) {
                biggest_area_rec = max(biggest_area_rec, rect_area(tile, tile2));
            }
        }
    }
    
    printf("Part 2: %"PRIu64"\n", biggest_area_rec);
}


int main(int argc, char **argv) {
    if(argc != 2) {
        printf("Usage: %s file_name\n", argv[0]);
        return -1;
    }

    const char * file_path = argv[1];

    Nob_String_Builder input_sb = {0};
    nob_read_entire_file(file_path, &input_sb);
    Nob_String_View input = nob_sv_from_parts(input_sb.items, input_sb.count);
    input = nob_sv_trim(input);
    
    part1(input);
    part2(input);

    nob_sb_free(input_sb);

    return 0;
}